File: src/screens/HomeScreen.js
import React, { useState } from 'react';
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
  Alert,
} from 'react-native';
import { useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/native';
import { submitCheckin, hasCheckedInToday } from '../services/checkinService';
import { isWithinCampus } from '../services/locationService';
import { AppConstants } from '../utils/constants';

export default function HomeScreen() {
  const navigation = useNavigation();
  const device = useSelector((state) => state.app.device);
  const currentQuestion = useSelector((state) => state.app.currentQuestion);
  const deviceId = useSelector((state) => state.app.deviceId);
  const [isCheckingIn, setIsCheckingIn] = useState(false);

  const handleCheckin = async () => {
    if (!deviceId) return;

    setIsCheckingIn(true);
    try {
      // 위치 확인
      const withinCampus = await isWithinCampus();
      if (!withinCampus) {
        Alert.alert('알림', '캠퍼스 범위 내에 있지 않습니다.');
        return;
      }

      // 오늘 이미 출석했는지 확인
      const hasCheckedIn = await hasCheckedInToday(deviceId);
      if (hasCheckedIn) {
        Alert.alert('알림', '오늘 이미 출석했습니다.');
        return;
      }

      // 출석 체크
      await submitCheckin(deviceId);
      Alert.alert('성공', '출석 완료! +10 포인트');
    } catch (error) {
      Alert.alert('오류', `출석 실패: ${error.message}`);
    } finally {
      setIsCheckingIn(false);
    }
  };

  return (
    <ScrollView style={styles.container}>
      {/* 포인트 표시 */}
      <View style={styles.pointCard}>
        <Text style={styles.pointLabel}>내 포인트</Text>
        <Text style={styles.pointValue}>{device?.points || 0}P</Text>
      </View>

      {/* 출석하기 버튼 */}
      <TouchableOpacity
        style={[styles.checkinButton, isCheckingIn && styles.buttonDisabled]}
        onPress={handleCheckin}
        disabled={isCheckingIn}
      >
        <Text style={styles.checkinButtonText}>
          {isCheckingIn ? '출석 중...' : '출석하기 (+10P)'}
        </Text>
      </TouchableOpacity>

      {/* 현재 질문 */}
      {currentQuestion ? (
        <View style={styles.questionCard}>
          <Text style={styles.questionLabel}>현재 고민</Text>
          <Text style={styles.questionText}>{currentQuestion.text}</Text>
          <View style={styles.questionInfo}>
            <Text style={styles.questionOptions}>
              {currentQuestion.optionA} vs {currentQuestion.optionB}
            </Text>
            <Text style={styles.voteCount}>
              {currentQuestion.totalVotes || 0}명 투표
            </Text>
          </View>
          <TouchableOpacity
            style={styles.voteButton}
            onPress={() => navigation.navigate('Vote', { question: currentQuestion })}
          >
            <Text style={styles.voteButtonText}>투표하기</Text>
          </TouchableOpacity>
        </View>
      ) : (
        <View style={styles.emptyCard}>
          <Text style={styles.emptyIcon}>❓</Text>
          <Text style={styles.emptyText}>현재 진행 중인 고민이 없습니다</Text>
          <TouchableOpacity
            style={styles.createButton}
            onPress={() => navigation.navigate('QuestionCreate')}
          >
            <Text style={styles.createButtonText}>고민 작성하기</Text>
          </TouchableOpacity>
        </View>
      )}

      {/* 고민 작성하기 버튼 */}
      <TouchableOpacity
        style={styles.createQuestionButton}
        onPress={() => navigation.navigate('QuestionCreate')}
      >
        <Text style={styles.createQuestionButtonText}>+ 고민 작성하기</Text>
      </TouchableOpacity>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f5f5f5',
    padding: 16,
  },
  pointCard: {
    backgroundColor: '#fff',
    padding: 20,
    borderRadius: 12,
    marginBottom: 16,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  pointLabel: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  pointValue: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#0ea5e9',
  },
  checkinButton: {
    backgroundColor: '#10b981',
    padding: 16,
    borderRadius: 12,
    marginBottom: 24,
    alignItems: 'center',
  },
  buttonDisabled: {
    opacity: 0.6,
  },
  checkinButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
  questionCard: {
    backgroundColor: '#fff',
    padding: 20,
    borderRadius: 12,
    marginBottom: 16,
    elevation: 2,
  },
  questionLabel: {
    fontSize: 14,
    color: '#666',
    marginBottom: 8,
  },
  questionText: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 16,
  },
  questionInfo: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 16,
  },
  questionOptions: {
    fontSize: 16,
    color: '#0ea5e9',
  },
  voteCount: {
    fontSize: 14,
    color: '#666',
  },
  voteButton: {
    backgroundColor: '#0ea5e9',
    padding: 16,
    borderRadius: 8,
    alignItems: 'center',
  },
  voteButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
  emptyCard: {
    backgroundColor: '#fff',
    padding: 40,
    borderRadius: 12,
    marginBottom: 16,
    alignItems: 'center',
  },
  emptyIcon: {
    fontSize: 60,
    marginBottom: 16,
  },
  emptyText: {
    fontSize: 16,
    color: '#666',
    marginBottom: 20,
  },
  createButton: {
    backgroundColor: '#0ea5e9',
    paddingHorizontal: 24,
    paddingVertical: 12,
    borderRadius: 8,
  },
  createButtonText: {
    color: '#fff',
    fontSize: 16,
    fontWeight: 'bold',
  },
  createQuestionButton: {
    backgroundColor: '#fff',
    padding: 16,
    borderRadius: 12,
    alignItems: 'center',
    borderWidth: 2,
    borderColor: '#0ea5e9',
  },
  createQuestionButtonText: {
    color: '#0ea5e9',
    fontSize: 18,
    fontWeight: 'bold',
  },
});

File: lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../providers/app_provider.dart';
import '../services/checkin_service.dart';
import '../services/location_service.dart';
import 'vote_screen.dart';
import 'question_create_screen.dart';
import 'profile_screen.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  final CheckinService _checkinService = CheckinService();
  final LocationService _locationService = LocationService();
  bool _isCheckingIn = false;

  Future<void> _handleCheckin() async {
    final appProvider = Provider.of<AppProvider>(context, listen: false);
    final deviceId = appProvider.deviceId;

    if (deviceId == null) return;

    setState(() {
      _isCheckingIn = true;
    });

    try {
      // 위치 확인
      final isWithinCampus = await _locationService.isWithinCampus();
      if (!isWithinCampus) {
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(
              content: Text('캠퍼스 범위 내에 있지 않습니다.'),
              backgroundColor: Colors.red,
            ),
          );
        }
        return;
      }

      // 오늘 이미 출석했는지 확인
      final hasCheckedIn = await _checkinService.hasCheckedInToday(deviceId);
      if (hasCheckedIn) {
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(
              content: Text('오늘 이미 출석했습니다.'),
              backgroundColor: Colors.orange,
            ),
          );
        }
        return;
      }

      // 출석 체크
      await _checkinService.submitCheckin(deviceId);

      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
            content: Text('출석 완료! +10 포인트'),
            backgroundColor: Colors.green,
          ),
        );
      }
    } catch (e) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text('출석 실패: $e'),
            backgroundColor: Colors.red,
          ),
        );
      }
    } finally {
      setState(() {
        _isCheckingIn = false;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('즉결심판'),
        actions: [
          IconButton(
            icon: const Icon(Icons.person),
            onPressed: () {
              Navigator.push(
                context,
                MaterialPageRoute(builder: (_) => const ProfileScreen()),
              );
            },
          ),
        ],
      ),
      body: Consumer<AppProvider>(
        builder: (context, appProvider, child) {
          final device = appProvider.device;
          final question = appProvider.currentQuestion;

          return SingleChildScrollView(
            padding: const EdgeInsets.all(16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                // 포인트 표시
                Card(
                  child: Padding(
                    padding: const EdgeInsets.all(16),
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        const Text(
                          '내 포인트',
                          style: TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                        Text(
                          '${device?.points ?? 0}P',
                          style: const TextStyle(
                            fontSize: 24,
                            fontWeight: FontWeight.bold,
                            color: Colors.blue,
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
                const SizedBox(height: 20),

                // 출석하기 버튼
                ElevatedButton.icon(
                  onPressed: _isCheckingIn ? null : _handleCheckin,
                  icon: _isCheckingIn
                      ? const SizedBox(
                          width: 20,
                          height: 20,
                          child: CircularProgressIndicator(strokeWidth: 2),
                        )
                      : const Icon(Icons.check_circle),
                  label: Text(_isCheckingIn ? '출석 중...' : '출석하기 (+10P)'),
                  style: ElevatedButton.styleFrom(
                    padding: const EdgeInsets.symmetric(vertical: 16),
                    backgroundColor: Colors.green,
                    foregroundColor: Colors.white,
                  ),
                ),
                const SizedBox(height: 30),

                // 현재 질문 표시
                if (question != null) ...[
                  Card(
                    elevation: 4,
                    child: Padding(
                      padding: const EdgeInsets.all(20),
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          const Text(
                            '현재 고민',
                            style: TextStyle(
                              fontSize: 14,
                              color: Colors.grey,
                            ),
                          ),
                          const SizedBox(height: 8),
                          Text(
                            question.text,
                            style: const TextStyle(
                              fontSize: 20,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                          const SizedBox(height: 16),
                          Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: [
                              Text(
                                '${question.optionA} vs ${question.optionB}',
                                style: const TextStyle(
                                  fontSize: 16,
                                  color: Colors.blue,
                                ),
                              ),
                              Text(
                                '${question.totalVotes}명 투표',
                                style: const TextStyle(
                                  fontSize: 14,
                                  color: Colors.grey,
                                ),
                              ),
                            ],
                          ),
                        ],
                      ),
                    ),
                  ),
                  const SizedBox(height: 20),
                  ElevatedButton(
                    onPressed: () {
                      Navigator.push(
                        context,
                        MaterialPageRoute(
                          builder: (_) => VoteScreen(question: question),
                        ),
                      );
                    },
                    style: ElevatedButton.styleFrom(
                      padding: const EdgeInsets.symmetric(vertical: 16),
                      backgroundColor: Colors.blue,
                      foregroundColor: Colors.white,
                    ),
                    child: const Text(
                      '투표하기',
                      style: TextStyle(fontSize: 18),
                    ),
                  ),
                ] else ...[
                  Card(
                    child: Padding(
                      padding: const EdgeInsets.all(40),
                      child: Column(
                        children: [
                          const Icon(
                            Icons.question_mark,
                            size: 60,
                            color: Colors.grey,
                          ),
                          const SizedBox(height: 16),
                          const Text(
                            '현재 진행 중인 고민이 없습니다',
                            style: TextStyle(
                              fontSize: 16,
                              color: Colors.grey,
                            ),
                          ),
                          const SizedBox(height: 20),
                          ElevatedButton(
                            onPressed: () {
                              Navigator.push(
                                context,
                                MaterialPageRoute(
                                  builder: (_) => const QuestionCreateScreen(),
                                ),
                              );
                            },
                            child: const Text('고민 작성하기'),
                          ),
                        ],
                      ),
                    ),
                  ),
                ],

                const SizedBox(height: 20),

                // 다음 고민 버튼
                OutlinedButton(
                  onPressed: () {
                    appProvider.refreshQuestion();
                  },
                  child: const Text('다음 고민 보기'),
                ),
                const SizedBox(height: 20),

                // 고민 작성하기 버튼
                ElevatedButton.icon(
                  onPressed: () {
                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (_) => const QuestionCreateScreen(),
                      ),
                    );
                  },
                  icon: const Icon(Icons.add),
                  label: const Text('고민 작성하기'),
                  style: ElevatedButton.styleFrom(
                    padding: const EdgeInsets.symmetric(vertical: 16),
                  ),
                ),
              ],
            ),
          );
        },
      ),
    );
  }
}

